File: test/test_astar_se2.cpp
#include <gtest/gtest.h>

#include "erl_search_planning/astar.hpp"
#include "erl_search_planning/planning_se2.hpp"
// THE ABOVE TWO HEADERS MUST BE INCLUDED BEFORE THE FOLLOWING HEADERS
#include "erl_geometry/bresenham_2d.hpp"

using namespace erl::common;
using namespace erl::env;
using namespace erl::geometry;
using namespace erl::search_planning;

TEST(AStarSE2Test, PlanWithOri25Eps2) {

    const int kNumThetas = 25;
    const double kCollisionCheckDt = 0.05;
    const double kEps = 2;
    const long kMaxNumReachedGoals = -1;
    const long kMaxNumIterations = -1;
    const bool kReopenInconsistent = false;
    const bool kLog = false;

    Eigen::Vector3d metric_goal_coords(50, 1, 0);
    Eigen::Vector3d metric_goal_tolerance = Eigen::Vector3d(0.25, 0.25, 0.53);
    Eigen::Scalard terrain_cost(0.);
    Eigen::Vector2i map_shape = Eigen::Vector2i(1001, 1001);
    Eigen::Vector2d map_min = Eigen::Vector2d::Zero();
    Eigen::Vector2d map_max = Eigen::Vector2d::Constant(100.0);  // resolution = 0.1
    auto grid_map_info = std::make_shared<GridMapInfo2D>(map_shape, map_min, map_max);
    // the loaded matrix is a map with top left corner at (0, 0) and bottom right corner at (100, 100), x increasing to the right and y increasing down,
    // column-major order (along the y-axis)
    // the grid map is a map with top left corner at (0, 0) and bottom right corner at (100, 100), x increasing down and y increasing to the right,
    // row-major order (along the y-axis). So, we flatten the matrix directly.
    Eigen::VectorX8U data = LoadEigenMatrixFromTextFile<uint8_t>(GRID_MAP_DATA_FILE).reshaped(map_shape.prod(), 1);
    auto grid_map = std::make_shared<GridMapUnsigned2D>(grid_map_info, std::move(data));
    auto motion_primitives = LoadDdcMotionPrimitivesFromYaml(DDC_MOTION_PRIMITIVES_FILE);
    auto planning_interface = std::make_shared<PlanningSe2>(
        metric_goal_coords,
        metric_goal_tolerance,
        terrain_cost,
        kCollisionCheckDt,
        motion_primitives,
        grid_map,
        kNumThetas);
    Eigen::Vector3d metric_start_coords(10, 90, 0);
    auto result = AStar(metric_start_coords, planning_interface, kEps, kMaxNumReachedGoals, kMaxNumIterations, kReopenInconsistent, kLog).Plan();

    std::cout << "Path length: " << result->paths[0].size() << std::endl
              << "Path cost: " << result->path_costs[0] << std::endl
              << "Number of actions: " << result->action_ids[0].size() << std::endl
              << "Number of inconsistent nodes: " << result->inconsistent_list.size() << std::endl;
    EXPECT_EQ(result->path_costs[0], 95);
    EXPECT_EQ(result->action_ids[0].size(), 95);
    EXPECT_EQ(result->inconsistent_list.size(), 0);
}

TEST(AStarSE2Test, PlanWithOri25Eps1) {

    const int kNumThetas = 25;
    const double kCollisionCheckDt = 0.05;
    const double kEps = 1;
    const long kMaxNumReachedGoals = -1;
    const long kMaxNumIterations = -1;
    const bool kReopenInconsistent = false;
    const bool kLog = false;

    Eigen::Vector3d metric_goal_coords(50, 1, 0);
    Eigen::Vector3d metric_goal_tolerance = Eigen::Vector3d(0.25, 0.25, 0.53);
    Eigen::Scalard terminal_cost(0.);
    Eigen::Vector2i map_shape = Eigen::Vector2i(1001, 1001);
    Eigen::Vector2d map_min = Eigen::Vector2d::Zero();
    Eigen::Vector2d map_max = Eigen::Vector2d::Constant(100.0);  // resolution = 0.1
    auto grid_map_info = std::make_shared<GridMapInfo2D>(map_shape, map_min, map_max);
    Eigen::VectorX8U data = LoadEigenMatrixFromTextFile<uint8_t>(GRID_MAP_DATA_FILE).reshaped(map_shape.prod(), 1);
    auto grid_map = std::make_shared<GridMapUnsigned2D>(grid_map_info, std::move(data));
    auto motion_primitives = LoadDdcMotionPrimitivesFromYaml(DDC_MOTION_PRIMITIVES_FILE);
    auto planning_interface = std::make_shared<PlanningSe2>(
        metric_goal_coords,
        metric_goal_tolerance,
        terminal_cost,
        kCollisionCheckDt,
        motion_primitives,
        grid_map,
        kNumThetas);
    Eigen::Vector3d metric_start_coords(10, 90, 0);
    auto result = AStar(metric_start_coords, planning_interface, kEps, kMaxNumReachedGoals, kMaxNumIterations, kReopenInconsistent, kLog).Plan();

    std::cout << "Path length: " << result->paths[0].size() << std::endl
              << "Path cost: " << result->path_costs[0] << std::endl
              << "Number of actions: " << result->action_ids[0].size() << std::endl
              << "Number of inconsistent nodes: " << result->inconsistent_list.size() << std::endl;
    EXPECT_EQ(result->path_costs[0], 95);
    EXPECT_EQ(result->action_ids[0].size(), 95);
    EXPECT_EQ(result->inconsistent_list.size(), 0);
}

TEST(AStarSE2Test, PlanWithOri101Eps1) {

    const int kNumThetas = 101;
    const double kCollisionCheckDt = 0.05;
    const double kEps = 1;
    const long kMaxNumReachedGoals = -1;
    const long kMaxNumIterations = -1;
    const bool kReopenInconsistent = false;
    const bool kLog = false;

    Eigen::Vector3d metric_goal_coords(50, 1, 0);
    Eigen::Vector3d metric_goal_tolerance = Eigen::Vector3d(0.25, 0.25, 0.53);
    Eigen::Scalard terminal_cost(0.);
    Eigen::Vector2i map_shape = Eigen::Vector2i(1001, 1001);
    Eigen::Vector2d map_min = Eigen::Vector2d::Zero();
    Eigen::Vector2d map_max = Eigen::Vector2d::Constant(100.0);  // resolution = 0.1
    auto grid_map_info = std::make_shared<GridMapInfo2D>(map_shape, map_min, map_max);
    Eigen::VectorX8U data = LoadEigenMatrixFromTextFile<uint8_t>(GRID_MAP_DATA_FILE).reshaped(map_shape.prod(), 1);
    auto grid_map = std::make_shared<GridMapUnsigned2D>(grid_map_info, std::move(data));
    auto motion_primitives = LoadDdcMotionPrimitivesFromYaml(DDC_MOTION_PRIMITIVES_FILE);
    auto planning_interface = std::make_shared<PlanningSe2>(
        metric_goal_coords,
        metric_goal_tolerance,
        terminal_cost,
        kCollisionCheckDt,
        motion_primitives,
        grid_map,
        kNumThetas);
    Eigen::Vector3d metric_start_coords(10, 90, 0);
    auto result = AStar(metric_start_coords, planning_interface, kEps, kMaxNumReachedGoals, kMaxNumIterations, kReopenInconsistent, kLog).Plan();

    std::cout << "Path length: " << result->paths[0].size() << std::endl
              << "Path cost: " << result->path_costs[0] << std::endl
              << "Number of actions: " << result->action_ids[0].size() << std::endl
              << "Number of inconsistent nodes: " << result->inconsistent_list.size() << std::endl;
    EXPECT_EQ(result->path_costs[0], 95);
    EXPECT_EQ(result->action_ids[0].size(), 95);
    EXPECT_EQ(result->inconsistent_list.size(), 0);
}

TEST(AStarSE2Test, PlanWithOri361Eps1) {

    const int kNumThetas = 361;
    const double kCollisionCheckDt = 0.05;
    const double kEps = 1;
    const long kMaxNumReachedGoals = -1;
    const long kMaxNumIterations = -1;
    const bool kReopenInconsistent = false;
    const bool kLog = false;

    Eigen::Vector3d metric_goal_coords(50, 1, 0);
    Eigen::Vector3d metric_goal_tolerance = Eigen::Vector3d(0.25, 0.25, 0.53);
    Eigen::Scalard terminal_cost(0.);
    Eigen::Vector2i map_shape = Eigen::Vector2i(1001, 1001);
    Eigen::Vector2d map_min = Eigen::Vector2d::Zero();
    Eigen::Vector2d map_max = Eigen::Vector2d::Constant(100.0);  // resolution = 0.1
    auto grid_map_info = std::make_shared<GridMapInfo2D>(map_shape, map_min, map_max);
    Eigen::VectorX8U data = LoadEigenMatrixFromTextFile<uint8_t>(GRID_MAP_DATA_FILE).reshaped(map_shape.prod(), 1);
    auto grid_map = std::make_shared<GridMapUnsigned2D>(grid_map_info, std::move(data));
    auto motion_primitives = LoadDdcMotionPrimitivesFromYaml(DDC_MOTION_PRIMITIVES_FILE);
    auto planning_interface = std::make_shared<PlanningSe2>(
        metric_goal_coords,
        metric_goal_tolerance,
        terminal_cost,
        kCollisionCheckDt,
        motion_primitives,
        grid_map,
        kNumThetas);
    Eigen::Vector3d metric_start_coords(10, 90, 0);
    auto result = AStar(metric_start_coords, planning_interface, kEps, kMaxNumReachedGoals, kMaxNumIterations, kReopenInconsistent, kLog).Plan();

    std::cout << "Path length: " << result->paths[0].size() << std::endl
              << "Path cost: " << result->path_costs[0] << std::endl
              << "Number of actions: " << result->action_ids[0].size() << std::endl
              << "Number of inconsistent nodes: " << result->inconsistent_list.size() << std::endl;
    EXPECT_EQ(result->path_costs[0], 95);
    EXPECT_EQ(result->action_ids[0].size(), 95);
    EXPECT_EQ(result->inconsistent_list.size(), 0);
}

TEST(AStarSE2Test, PlanWithTriangleShape) {

    const int kNumThetas = 25;
    const double kCollisionCheckDt = 0.05;
    const double kEps = 1;
    const long kMaxNumReachedGoals = -1;
    const long kMaxNumIterations = -1;
    const bool kReopenInconsistent = false;
    const bool kLog = false;

    Eigen::Vector3d metric_goal_coords(50, 1, 0);
    Eigen::Vector3d metric_goal_tolerance = Eigen::Vector3d(0.25, 0.25, 0.53);
    Eigen::Scalard terminal_cost(0.);
    Eigen::Vector2i map_shape = Eigen::Vector2i(1001, 1001);
    Eigen::Vector2d map_min = Eigen::Vector2d::Zero();
    Eigen::Vector2d map_max = Eigen::Vector2d::Constant(100.0);  // resolution = 0.1
    auto grid_map_info = std::make_shared<GridMapInfo2D>(map_shape, map_min, map_max);
    Eigen::VectorX8U data = LoadEigenMatrixFromTextFile<uint8_t>(GRID_MAP_DATA_FILE).reshaped(map_shape.prod(), 1);
    auto grid_map = std::make_shared<GridMapUnsigned2D>(grid_map_info, std::move(data));
    auto motion_primitives = LoadDdcMotionPrimitivesFromYaml(DDC_MOTION_PRIMITIVES_FILE);
    Eigen::Matrix23d triangle_vertices;
    // clang-format off
    triangle_vertices << -0.3, 0.3, -0.3,
                         -0.2,   0,  0.2;
    // clang-format on
    triangle_vertices.array() *= 3;
    // Eigen::Matrix2Xd metric_grids = grid_map_info->GetMetricCoordinatesOfFilledMetricPolygon(triangle_vertices);
    // If we only use the contour, we may miss some obstacles that are inside the triangle.
    // Eigen::Matrix2Xd metric_grids =
    //     grid_map_info->PixelToMeterForPoints(ComputePixelsOfPolygonContour(grid_map_info->MeterToPixelForPoints(triangle_vertices)));
    double inflate_scale = 1.0;
    auto planning_interface = std::make_shared<PlanningSe2>(
        metric_goal_coords,
        metric_goal_tolerance,
        terminal_cost,
        kCollisionCheckDt,
        motion_primitives,
        grid_map,
        kNumThetas,
        inflate_scale,
        triangle_vertices);
    Eigen::Vector3d metric_start_coords(10, 90, 0);
    auto result = AStar(metric_start_coords, planning_interface, kEps, kMaxNumReachedGoals, kMaxNumIterations, kReopenInconsistent, kLog).Plan();

    std::cout << "Path length: " << result->paths[0].size() << std::endl
              << "Path cost: " << result->path_costs[0] << std::endl
              << "Number of actions: " << result->action_ids[0].size() << std::endl
              << "Number of inconsistent nodes: " << result->inconsistent_list.size() << std::endl;
    EXPECT_EQ(result->path_costs[0], 95);
    EXPECT_EQ(result->action_ids[0].size(), 95);
    EXPECT_EQ(result->inconsistent_list.size(), 0);
}

File: test/test_astar_se2.py
import time
import unittest

import matplotlib as mpl
import matplotlib.pyplot as plt
import numpy as np
from erl_path_planning.common.storage import GridMapInfo2D
from erl_path_planning.common.storage import GridMapUnsigned2D
from erl_path_planning.env import load_ddc_motion_primitives_from_yaml
from erl_path_planning import AStar
from erl_path_planning import PlanningSe2
from erl_path_planning.env.data.house_expo import HouseExpoMap
from erl_path_planning.env.data.house_expo.list_data import get_map_and_traj_files


class TestAStarSe2(unittest.TestCase):
    # def test_plan_with_point_collision_check(self):
    #     num_orientations = 25
    #     collision_check_dt = 0.05
    #     metric_start_coords = np.array([10.0, 90.0, 0.0])
    #     metric_goal_coords = np.array([50.0, 1.0, 0.0])
    #     metric_goal_tolerance = np.array([0.25, 0.25, 0.25])
    #     terminal_cost = np.array([0.0])
    #     map_shape = np.array([1001, 1001])
    #     map_min = np.array([0.0, 0.0])
    #     map_max = np.array([100.0, 100.0])
    #     grid_map_info = GridMapInfo2D(map_shape, map_min, map_max)
    #     data = np.loadtxt(
    #         "circles_map_1001x1001.txt", dtype=np.uint8, delimiter=" "
    #     )  # x to the right, y to the bottom, row-major (along x first)
    #     grid_map = GridMapUnsigned2D(
    #         grid_map_info, data.transpose().flatten()
    #     )  # x to the bottom, y to the right, row-major (along y first)
    #     motion_primitives = load_ddc_motion_primitives_from_yaml("ddc_motion_primitives.yaml")
    #     planning_se2 = PlanningSe2(
    #         metric_goals_coords=metric_goal_coords,
    #         metric_goals_tolerance=metric_goal_tolerance,
    #         terminal_costs=terminal_cost,
    #         collision_check_dt=collision_check_dt,
    #         motion_primitives=motion_primitives,
    #         grid_map=grid_map,
    #         num_thetas=num_orientations,
    #     )
    #     result = AStar(
    #         metric_start_coords,
    #         planning_se2,
    #         eps=2,
    #         max_num_reached_goals=1,
    #         max_num_iterations=-1,
    #         reopen_inconsistent=False,
    #         log=False,
    #     ).plan()
    #     self.assertEqual(len(result.paths[0]), 97)
    #     self.assertEqual(result.path_costs[0], 96)
    #     self.assertEqual(len(result.action_ids[0]), 96)
    #     self.assertEqual(len(result.inconsistent_list), 0)
    #
    # def test_plan_with_triangle_shape(self):
    #     num_orientations = 25
    #     collision_check_dt = 0.05
    #     metric_start_coords = np.array([10.0, 90.0, 0.0])
    #     metric_goal_coords = np.array([94.0, 7.0, 0.0])
    #     metric_goal_tolerance = np.array([0.25, 0.25, 0.53])
    #     terminal_cost = np.array([0.0])
    #     map_shape = np.array([1001, 1001])
    #     map_min = np.array([0.0, 0.0])
    #     map_max = np.array([100.0, 100.0])
    #     grid_map_info = GridMapInfo2D(map_shape, map_min, map_max)
    #     data = np.loadtxt(
    #         "circles_map_1001x1001.txt", dtype=np.uint8, delimiter=" "
    #     )  # x to the right, y to the bottom, row-major (along x first)
    #     grid_map = GridMapUnsigned2D(
    #         grid_map_info, data.transpose().flatten()
    #     )  # x to the bottom, y to the right, row-major (along y first)
    #     motion_primitives = load_ddc_motion_primitives_from_yaml("ddc_motion_primitives.yaml")
    #     triangle_vertices = np.array([[-0.3, 0.3, -0.3], [-0.2, 0.0, 0.2]]) * 3
    #     # metric_grids = grid_map_info.get_metric_coordinates_of_filled_metric_polygon(triangle_vertices)
    #     # metric_grids = grid_map_info.cell_to_meter_for_points(compute_pixels_of_polygon_contour(grid_map_info.meter_to_cell_for_points(triangle_vertices)))
    #     planning_se2 = PlanningSe2(
    #         metric_goals_coords=metric_goal_coords,
    #         metric_goals_tolerance=metric_goal_tolerance,
    #         terminal_costs=terminal_cost,
    #         collision_check_dt=collision_check_dt,
    #         motion_primitives=motion_primitives,
    #         grid_map=grid_map,
    #         num_thetas=num_orientations,
    #         inflate_scale=1.0,
    #         shape_metric_vertices=triangle_vertices,
    #     )
    #     result = AStar(
    #         metric_start_coords,
    #         planning_se2,
    #         eps=2,
    #         max_num_reached_goals=1,
    #         max_num_iterations=-1,
    #         reopen_inconsistent=False,
    #         log=False,
    #     ).plan()
    #
    #     plt.figure(figsize=(10, 10))
    #     extend = np.array([map_min[0], map_max[0], map_min[1], map_max[1]])
    #     plt.imshow(data[::-1], cmap="binary_r", extent=extend)  # y to the bottom -> y to the top
    #
    #     path = np.array(result.paths[0])  # rough path
    #     np.savetxt("path.txt", path)
    #     plt.plot(path[:, 0], path[:, 1], "r-")
    #
    #     body_patch = plt.Polygon(triangle_vertices.transpose(), closed=False, color="b", fill=True)
    #     plt.gca().add_patch(body_patch)
    #
    #     def update_body_patch(x, y, theta):
    #         s = np.sin(theta)
    #         c = np.cos(theta)
    #         rot = np.array([[c, -s], [s, c]])
    #         body_patch.set_xy((rot @ triangle_vertices + np.array([[x], [y]])).T)
    #
    #     update_body_patch(metric_start_coords[0], metric_start_coords[1], metric_start_coords[2])
    #
    #     path_line = plt.plot(metric_start_coords[0], metric_start_coords[1], "g-")[0]
    #
    #     def update_path_line(x, y):
    #         path_line.set_data(x, y)
    #
    #     state = metric_start_coords
    #     path = []
    #     for action_id in result.action_ids[0]:
    #         trajectory = planning_se2.forward_action_in_metric_space(state, action_id, 0.01).T
    #
    #         for metric_state in trajectory:
    #             update_body_patch(metric_state[0], metric_state[1], metric_state[2])
    #             path.append(metric_state)
    #             update_path_line(np.array(path)[:, 0], np.array(path)[:, 1])
    #
    #         state = trajectory[-1]
    #
    #         plt.pause(0.1)
    #     plt.show(block=False)
    #     plt.pause(5)

    def test_plan_with_house_expo_map(self):
        map_file = get_map_and_traj_files()[1451][0]
        # house_expo_map = HouseExpoMap(map_file, 0.15)
        house_expo_map = HouseExpoMap(map_file)
        vertices = house_expo_map.meter_space.surface.vertices
        print(vertices.shape)
        map_min = np.min(vertices, axis=1)
        map_max = np.max(vertices, axis=1)
        grid_map_info = GridMapInfo2D(
            min=map_min,
            max=map_max,
            resolution=np.array([0.05, 0.05]),
            padding=np.array([4, 4]),
        )
        map_image = house_expo_map.meter_space.generate_map_image(grid_map_info)
        print(map_image.flags["C_CONTIGUOUS"])
        print(map_image.flags["F_CONTIGUOUS"])
        print(map_image.min(), map_image.max(), flush=True)
        grid_map = GridMapUnsigned2D(grid_map_info, 255 - map_image[::-1].T.flatten())
        motion_primitives = load_ddc_motion_primitives_from_yaml("ddc_motion_primitives.yaml")
        triangle_vertices = np.array([[-0.3, 0.3, -0.3], [-0.2, 0.0, 0.2]]) * 0.5
        # metric_grids = grid_map_info.get_metric_coordinates_of_filled_metric_polygon(triangle_vertices)
        # metric_grids = grid_map_info.cell_to_meter_for_points(compute_pixels_of_polygon_contour(grid_map_info.meter_to_cell_for_points(triangle_vertices)))
        num_orientations = 91
        collision_check_dt = 0.01
        metric_start_coords = np.array([7.0, 7.0, 0.0])
        metric_goal_coords = np.array([10.0, 1.0, 3.1])
        metric_goal_tolerance = np.array([0.25, 0.25, 0.25])
        terminal_cost = np.array([0.0])
        n = 10
        t_total = 0.0
        for _ in range(n):
            planning_se2 = PlanningSe2(
                metric_goals_coords=metric_goal_coords,
                metric_goals_tolerance=metric_goal_tolerance,
                terminal_costs=terminal_cost,
                collision_check_dt=collision_check_dt,
                motion_primitives=motion_primitives,
                grid_map=grid_map,
                num_thetas=num_orientations,
                inflate_scale=1.5,
                shape_metric_vertices=triangle_vertices,
            )
            t0 = time.time()
            result = AStar(
                metric_start_coords,
                planning_se2,
                eps=4,
                max_num_reached_goals=1,
                max_num_iterations=-1,
                reopen_inconsistent=False,
                log=False,
            ).plan()
            t_total += time.time() - t0
        print("time: ", t_total / n, flush=True)

        plt.figure(figsize=(10, 10))
        extend = np.array([map_min[0], map_max[0], map_min[1], map_max[1]])
        plt.imshow(map_image, cmap="binary_r", extent=extend)  # y to the bottom -> y to the top
        # plt.show()

        np.set_printoptions(linewidth=10000000)
        print(result.paths[0])

        path = result.paths[0].T[::-10][::-1]  # rough path
        plt.plot(path[:, 0], path[:, 1], "r-")

        body_patch = plt.Polygon(triangle_vertices.transpose(), closed=False, color="b", fill=True)
        plt.gca().add_patch(body_patch)

        def update_body_patch(x, y, theta):
            s = np.sin(theta)
            c = np.cos(theta)
            rot = np.array([[c, -s], [s, c]])
            body_patch.set_xy((rot @ triangle_vertices + np.array([[x], [y]])).T)

        update_body_patch(metric_start_coords[0], metric_start_coords[1], metric_start_coords[2])

        path_line = plt.plot(metric_start_coords[0], metric_start_coords[1], "g-")[0]

        def update_path_line(x, y):
            path_line.set_data(x, y)

        for i in range(path.shape[0]):
            update_body_patch(path[i, 0], path[i, 1], path[i, 2])
            update_path_line(path[: i + 1, 0], path[: i + 1, 1])
            plt.pause(0.1)

        plt.show(block=False)
        plt.pause(5)


if __name__ == "__main__":
    unittest.main()
